core/perf_test/PerfTest_CustomReduction.cpp

```diff
@@ -0,0 +1,129 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact  H. Carter Edwards (hcedwar@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <Kokkos_Core.hpp>
+#include <gtest/gtest.h>
+#include <PerfTest_Category.hpp>
+#include <Kokkos_Random.hpp>
+
+namespace Test {
+template<class Scalar>
+void custom_reduction_test(int N, int R, int num_trials) {
+  Kokkos::Random_XorShift64_Pool<> rand_pool(183291);
+  Kokkos::View<Scalar*> a("A",N);
+  Kokkos::fill_random(a,rand_pool,1.0);
+
+  Scalar max;
+
+  // Warm up
+  Kokkos::parallel_reduce(Kokkos::TeamPolicy<>(N/1024,32), KOKKOS_LAMBDA( const Kokkos::TeamPolicy<>::member_type& team, Scalar& lmax) {
+    Scalar team_max = Scalar(0);
+    for(int rr = 0; rr<R; rr++) {
+    int i = team.league_rank();
+    Kokkos::parallel_reduce(Kokkos::TeamThreadRange(team,32), [&] (const int& j, Scalar& thread_max) {
+      Scalar t_max = Scalar(0);
+      Kokkos::parallel_reduce(Kokkos::ThreadVectorRange(team,32), [&] (const int& k, Scalar& max_) {
+        const Scalar val =  a((i*32 + j)*32 + k);
+        if(val>lmax) lmax = val;
+        if((k == 11) && (j==17) && (i==2)) lmax = 11.5;
+#ifdef GENERIC_REDUCER
+      },Kokkos::Max<Scalar>(t_max));
+#else
+      },Kokkos::Experimental::Max<Scalar>(t_max));
+#endif
+      if(t_max>thread_max) thread_max = t_max;
+#ifdef GENERIC_REDUCER
+    },Kokkos::Max<Scalar>(team_max));
+#else
+    },Kokkos::Experimental::Max<Scalar>(team_max));
+#endif
+    }
+    if(team_max>lmax) lmax = team_max;
+  },Kokkos::Experimental::Max<Scalar>(max));
+
+  // Timing
+  Kokkos::Timer timer;
+  for(int r = 0; r<num_trials; r++) {
+    Kokkos::parallel_reduce(Kokkos::TeamPolicy<>(N/1024,32), KOKKOS_LAMBDA( const Kokkos::TeamPolicy<>::member_type& team, Scalar& lmax) {
+      Scalar team_max = Scalar(0);
+      for(int rr = 0; rr<R; rr++) {
+      int i = team.league_rank();
+      Kokkos::parallel_reduce(Kokkos::TeamThreadRange(team,32), [&] (const int& j, Scalar& thread_max) {
+        Scalar t_max = Scalar(0);
+        Kokkos::parallel_reduce(Kokkos::ThreadVectorRange(team,32), [&] (const int& k, Scalar& max_) {
+          const Scalar val =  a((i*32 + j)*32 + k);
+          if(val>lmax) lmax = val;
+          if((k == 11) && (j==17) && (i==2)) lmax = 11.5;
+#ifdef GENERIC_REDUCER
```

Comment:
No this goes away. Was just there for testing, I am getting rid of the impl/Kokkos_Reducer.hpp